Bugfix: client hangs after run container

WeiZhang555 opened the pull request:
Fixes https://github.com/docker/docker/issues/25538 
Fixes https://github.com/docker/docker/issues/25533

Fixes issue introduced by https://github.com/docker/docker/pull/20848

New codes introduced a new method to wait container's exit code and
removal via Events API, but it specified a "since" filter based on
client side clock, which sometimes isn't synced up with daemon's clock,
hence leads the client to hang on waiting container's exit status.

This commit brings the Events call before start, and removes the time
filter, so that it can catch daemon events correctly without care of the
clock issue.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

WeiZhang555 commented:
Mirror PR: #25513

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "debug-win2lin" git@github.com:WeiZhang555/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

WeiZhang555 commented:
https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/29427/console

Looks like it's stuck on DockerSuite.TestBuildBuildTimeArg

WeiZhang555 committed:
Bugfix: client hangs after run container

New codes introduced a new method to wait container's exit code and
removal via Events API, but it specified a "since" filter based on
client side clock, which sometimes isn't synced up with daemon's clock,
hence leads the client to hang on waiting container's exit status.

This commit brings the Events call before start, and removes the time
filter, so that it can catch daemon events correctly without care of the
clock issue.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

thaJeztah commented:
Got a panic in another build-time arg test (`TestBuildBuildTimeArgOverrideEnvDefinedBeforeArg`) here: https://github.com/docker/docker/pull/25406#issuecomment-238516472

WeiZhang555 commented:
@thaJeztah @tonistiigi 

I think I find the root cause, it's caused by here: https://github.com/docker/docker/blob/master/api/client/container/run.go#L147

I use a "startTime" as a "since" filter of "Events" API, then the container rely on the Events API to determine when to quit. But, if a client visit a remote daemon, things will go wrong. They often don't have exactly same time!

I'm doing a quick fix, I think I can make it today.

WeiZhang555 commented:
That can also explain why sometime my PR passed the `win2lin` tests but sometime not, the windows client didn't sync up the time with remote linux daemon.

michael-holzheu commented:
As reported in #25533:

@WeiZhang555 unfortunately the commit does not fix the problem on s390x:
```
# git checkout origin/pr/25524
HEAD is now at 8541587... Bugfix: client hangs after run container
# TESTFLAGS='-check.f DockerSuite.TestEventsLimit'  \
       DOCKER_GRAPHDRIVER=vfs DOCKER_EXECDRIVER=native TIMEOUT="120m" \
       hack/make.sh gccgo test-integration-cli
+ go test -check.f DockerSuite.TestEventsLimit -check.v -check.timeout=120m -test.timeout=360m github.com/docker/docker/integration-cli
INFO: Testing against a local daemon
Cannot connect to the Docker daemon. Is the docker daemon running on this host?

Cannot connect to the Docker daemon. Is the docker daemon running on this host?
----------------------------------------------------------------------
FAIL: docker_cli_events_test.go:108: TestEventsLimit.pN52_github_com_docker_docker_integration_cli.DockerSuite

docker_cli_events_test.go:129:
    c.Assert(err, checker.IsNil, check.Commentf("%q failed with error", strings.Join(args, " ")))
... value *errors.errorString = &errors.errorString{s:"exit status 255: time=\"2016-08-09T16:24:04Z\" level=error msg=\"decode events error: unexpected EOF\" \n"} ("exit status 255: time=\"2016-08-09T16:24:04Z\" level=error msg=\"decode events error: unexpected EOF\" \n")
... "run --rm busybox true" failed with error

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (14.43s)
FAIL
---> Making bundle: .integration-daemon-stop (in bundles/1.13.0-dev/test-integration-cli)
+++++ cat bundles/1.13.0-dev/test-integration-cli/docker.pid
++++ kill 25560
/go/src/github.com/docker/docker/hack/make/.integration-daemon-stop: line 8: kill: (25560) - No such process
```

WeiZhang555 commented:
@michael-holzheu I don't have a  s390x hardware, so it will be a little bit hard for me to debug. But I'll try to find a fix for you.

At least this fix works for win2lin, and integration-test also shows that it works fine.

WeiZhang555 commented:
I have to sleep now. Tomorrow I will do more fix.

justincormack commented:
@WeiZhang555 shall we merge this if it looks good anyway to fix win2lin? Sleep well!

mlaventure commented:
@michael-holzheu could you put the daemon logs for that test somewhere for us to look at?

michael-holzheu commented:
> @michael-holzheu could you put the daemon logs for that test somewhere for us to look at?

 @mlaventure I checked test-integration-cli/docker.log and found the following:

```
panic: runtime error: index out of range

goroutine 361 [running]:

github_com_tchap_go_patricia_patricia.remove.pN53_github_com_tchap_go_patricia_patricia.sparseChildList
        /go/src/github.com/docker/docker/vendor/src/github.com/tchap/go-patricia/patricia/children.go:73
github_com_tchap_go_patricia_patricia.Delete.pN42_github_com_tchap_go_patricia_patricia.Trie
        /go/src/github.com/docker/docker/vendor/src/github.com/tchap/go-patricia/patricia/patricia.go:302
github_com_docker_docker_pkg_truncindex.Delete.pN50_github_com_docker_docker_pkg_truncindex.TruncIndex
        /go/src/github.com/docker/docker/pkg/truncindex/truncindex.go:95
```

The reason for this is that the PR #25524 branch does not contain 3d714b5ed58cfdfd5872ddd3654d171b09bb02d3. Sorry for missing this.

After I rebased the branch __it looks good__:

```
# TESTFLAGS='-check.f DockerSuite.TestEventsLimit'        DOCKER_GRAPHDRIVER=vfs DOCKER_EXECDRIVER=native TIMEOUT="120m"       hack/make.sh gccgo test-integration-cli
+ go test -check.f DockerSuite.TestEventsLimit -check.v -check.timeout=120m -test.timeout=360m github.com/d
INFO: Testing against a local daemon
PASS: docker_cli_events_test.go:108: TestEventsLimit.pN52_github_com_docker_docker_integration_cli.DockerSuite     15.288s
OK: 1 passed
PASS
```

WeiZhang555 commented:
Let me re-trigger the tests to prove it works, for the peace of myself.

@justincormack Yes, this is ready to be merged to fix win2lin, this can be taken as a formal fix.

/cc @mlaventure @tonistiigi @cpuguy83 for review, sorry for the inconvenience I brought to everyone :-(

WeiZhang555 commented:
All green now! :tada:

thaJeztah commented:
Thanks @WeiZhang555, I'll leave review to the others, as I'm not very good at this stuff :blush:

LK4D4 commented:
Ok, I hope we can figure this out here. There are at least two races in current implementation(they were there before this PR as well):
```
WARNING: DATA RACE
Read by goroutine 21:
  github.com/docker/docker/api/client/container.waitExitOrRemoved.func5()
      /home/moroz/project/workspace/src/github.com/docker/docker/api/client/container/utils.go:106 +0x1b7

Previous write by goroutine 27:
  github.com/docker/docker/api/client/container.waitExitOrRemoved.func1()
      /home/moroz/project/workspace/src/github.com/docker/docker/api/client/container/utils.go:38 +0x1e9
```
```
WARNING: DATA RACE                                                                                                                                                                     [16/218]
Read by goroutine 20:
  github.com/docker/engine-api/client/transport/cancellable.(*notifyingReader).Read()
      /home/moroz/project/workspace/src/github.com/docker/docker/vendor/src/github.com/docker/engine-api/client/transport/cancellable/cancellable.go:99 +0xcb
  encoding/json.(*Decoder).refill()
      /home/moroz/go/src/encoding/json/stream.go:152 +0x47e
  encoding/json.(*Decoder).readValue()
      /home/moroz/go/src/encoding/json/stream.go:128 +0x6a5
  encoding/json.(*Decoder).Decode()
      /home/moroz/go/src/encoding/json/stream.go:57 +0x1ef
  github.com/docker/docker/api/client/system.DecodeEvents()
      /home/moroz/project/workspace/src/github.com/docker/docker/api/client/system/events_utils.go:56 +0x15f
  github.com/docker/docker/api/client/container.waitExitOrRemoved.func4()
      /home/moroz/project/workspace/src/github.com/docker/docker/api/client/container/utils.go:78 +0xd1

Previous write by goroutine 21:
  github.com/docker/engine-api/client/transport/cancellable.(*notifyingReader).Close()
      /home/moroz/project/workspace/src/github.com/docker/docker/vendor/src/github.com/docker/engine-api/client/transport/cancellable/cancellable.go:110 +0xd2
  github.com/docker/docker/api/client/container.waitExitOrRemoved.func5()
      /home/moroz/project/workspace/src/github.com/docker/docker/api/client/container/utils.go:108 +0x208
```
They're related and are the source of panic.
@WeiZhang555 Let me know your thoughts.

LK4D4 commented:
The second panic should be fixed by docker/engine-api#357

vdemeester commented:
docker/engine-api#357 is merged :angel:

WeiZhang555 commented:
@LK4D4 @vdemeester 

The second race can cause  panic due to closing channel twice, as to the first race, I don't see the damage it will bring. 
The WARNING of first race is complaining that one goroutine is reading the exitCode variable, the other goroutine is writting this parameter, but I don't think it will bring problem, it won't panic either. 

However I can add a lock to protect the `exitCode` (https://github.com/docker/docker/pull/25524/files#diff-e905cdb1629f5a61235a6608666ca3a6L25), but I don't think it's necessary, is it?

What do you think of this? @vdemeester

WeiZhang555 commented:
And how did you find the issue?  Can you tell me your method? Or I can verify when I'm trying to fix this race. @LK4D4  Thanks. :smile:

WeiZhang555 commented:
I added a `atomic` for protecting the `exitCode`, the warning is expected to be gone whatever.

cpuguy83 commented:
This `waitExitOrRemoved` ~~has gotten~~ is quite unwieldy. I would prefer if we can reduce some of the complexity here rather than adding more.

cpuguy83 commented on api/client/container/utils.go:
```diff
@@ -54,45 +57,59 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
+			case waitErr = <-eventsErr:
+				atomic.StoreInt32(&exitCode, 125)
+			}
+		} else {
```
Seems we can get rid of this else since `destroyChan` since it won't hurt to always be listening for a notification on this channel.

cpuguy83 commented on api/client/container/utils.go:
```diff
@@ -54,45 +57,59 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
+			}
+		} else {
+			select {
+			case <-exitChan:
```
`exitChan` and `destroyChan` are basically the same thing.
Maybe we can combine these to a single `doneChan`

WeiZhang555 replied:
Yep, good idea. Actually destroyChan, exitChan and detachChan can be combined into one.

cpuguy83 commented on api/client/container/utils.go:
```diff
@@ -16,26 +16,29 @@ import (
+					if code, err := strconv.Atoi(v); err != nil {
 						logrus.Errorf("Can't convert %q to int: %v", v, err)
+					} else {
+						atomic.StoreInt32(&exitCode, int32(code))
```
Why not send the exitcode down the channel instead of just closing it?

WeiZhang555 replied:
Because channel getting data indicates the wait is finished, but indeed we still need to go on waiting for container's removal.

cpuguy83 commented:
Specifically I think it would be easier to reason about if instead of returning a channel, let the caller orchestrate that.

michael-holzheu commented:
@WeiZhang555 and others: Any outlook when this will be fixed? This currently blocks our CI on s390x.

mlaventure commented:
@WeiZhang555 @cpuguy83 what about something like https://gist.github.com/mlaventure/5d3cc5a97c9af83d65e6c202f086d7df ?

cpuguy83 commented:
@mlaventure Definitely better.

WeiZhang555 commented:
I think @mlaventure 's function is good, only with some small nit. 
Do you mind that I copy your code? Or maybe you can make a PR to carry this one. :smile:

WeiZhang555 commented:
@mlaventure Most of the codes are good, only this line https://gist.github.com/mlaventure/5d3cc5a97c9af83d65e6c202f086d7df#file-waitexitorremoved-go-L23,
I think maybe you forget to set the exitCode? Missing 
`exitCode=code` something like this. :smile:

WeiZhang555 commented:
@mlaventure I'll copy your code and sign you name, hope you don't mind :smile:

mlaventure commented:
@WeiZhang555 go ahead, you can copy it. Sorry, I didn't test it before posting, it was just for the idea.

Just did though (after fixing the compile error), and it passed the tests that you wrote with your PR

WeiZhang555 committed:
Refactor waitExitOrRemoved by mlaventure

Reduce complexity of waitExitOrRemoved.

Signed-off-by: mlaventure
Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

mlaventure commented:
@WeiZhang555 `api/client/container/utils.go:39: missing argument for Errorf("%v"): format reads arg 2, have only 1 args`

forgot to add `v` before `cerr` :sweat_smile:

WeiZhang555 commented:
@mlaventure Yep, I noticed, already fixed. 

It's late here, I think I have to go to bed later. Since the bug blocks a lot of CIs, feel free to carry this if you think it's necessary to make the process faster.  Or I can handle it tomorrow morning.
:smile:

mlaventure commented:
`DockerSuite.TestPluginForceRemove` failed on experimental, but it's a known issue with plugin (see #25511).

So, LGTM.

LK4D4 commented:
@WeiZhang555 I've built docker with `-race` flag to `go install` and run `docker run -it --rm busybox true`. :)

cpuguy83 commented:
LGTM
